### DISCUSSION TITLE
Highly experimental UI dataflow lib

### DISCUSSION THREAD

**Discussion by TheRawMeatball** [bevyengine/bevy#2882] [Show and tell]
After many attempts, I finally have a UI lib that's done enough to be tested by the public:
https://github.com/TheRawMeatball/ui4 (needs https://github.com/TheRawMeatball/bevy/tree/runs-ui4 to run)
This is *not* a release, there's still much polishing to do, and there's still major unimplemented features. However, what exists right now (in my opinion) is still very interesting and potentially useful.

If you want to help, just checking out the example, tinkering with it and sharing what you make would be incredibly helpful. Implementing more widgets, particularly a textbox and a checkbox in the style of widgets already present in the example would be even more helpful.

For general discussion / asking any sort of questions about how things work the #ui-dev channel on the bevy discord is preffered, but this discussion will also work :)

![Peek 2021-09-28 00-29](https://user-images.githubusercontent.com/6781733/134988097-e8d1c61f-3a3c-403d-8231-c488a6f6af6f.gif)

**Comment by afonsolage** [comment]
First of all: Amazing! I love the general idea and I think this is a very nice approach. I haven't read the source yet, so I'll give some feedback based on example, of course as you stated, some of those points may due to it's early stage in development, but I think worth pointing anyway:

1. `let this = ctx.this();`
I didn't understood what this does. I can imagine it's some kind of clone self reference, but I think this should have a more meaningful name


2. 
```
        .add_startup_system(
            (|world: &mut World| init_ui(world, root))
                .exclusive_system()
                .at_end(),
        )
```
Is it possible to mess around with system trait and do something like: `.add_startup_system(ui4::init_ui.with_root(my_root))` 

3.
```
            .with(HoverFunc(ButtonFunc::new(move |world| {
                world
                    .entity_mut(this)
                    .get_mut::<ButtonHoverState>()
                    .unwrap()
                    .0 = true;
            })))
```
This means all UI reated functions/callbacks runs within an exclusive system?

**Comment by TheRawMeatball** [comment]
> let this = ctx.this();

This just gets the `Entity` for the widget being built - I think it'll be clear enough when I add some docs :)

For the second point, something kinda like that is *probably* possible, I might even integrate it with the plugin, but right now the init API isn't the focus. I'll keep that in mind when I get back to it though. And for the third point, yes, callbacks will be ran from an exclusive system.
